pkg/image/prune/orphanedimagefinder.go

```diff
@@ -0,0 +1,162 @@
+	addPodsToGraph(g, pods, registryURL)
+	addReplicationControllersToGraph(g, rcs, registryURL)
+
+	return &orphanedImageFinder{g: g}
+}
+
+const minimumAgeInMinutesToPrune = 60
+
+func addImagesToGraph(g graph.Graph, images *imageapi.ImageList, registryURL string) {
+	for i := range images.Items {
+		image := &images.Items[i]
+		if !strings.HasPrefix(image.DockerImageReference, registryURL) {
+			continue
+		}
+		age := util.Now().Sub(image.CreationTimestamp.Time)
+		if age.Minutes() < minimumAgeInMinutesToPrune {
+			continue
+		}
+		graph.Image(g, image)
+	}
+}
+
+func addImageStreamsToGraph(g graph.Graph, streams *imageapi.ImageStreamList, registryURL string) {
+	for i := range streams.Items {
+		stream := &streams.Items[i]
+		if len(stream.Spec.DockerImageRepository) > 0 {
+			// skip streams that point at external registries
+			continue
+		}
+		imageStreamNode := graph.ImageStream(g, stream)
+		for _, history := range stream.Status.Tags {
+			if len(history.Items) == 0 {
+				continue
+			}
+			for i := range history.Items {
+				glog.Infof("Looking for node with image id %q", history.Items[i].Image)
+				imageNode := graph.FindImage(g, history.Items[i].Image)
+				glog.Infof("Found: %#v", imageNode)
+				if imageNode != nil {
+					var kind int
+					switch i {
+					case 0:
+						kind = graph.ReferencedImageGraphEdgeKind
+					default:
+						kind = graph.WeakReferencedImageGraphEdgeKind
+					}
+					g.AddEdge(imageStreamNode, imageNode, kind)
+				}
+			}
+		}
+	}
+}
+
+func addPodsToGraph(g graph.Graph, pods *kapi.PodList, registryURL string) {
+	for i := range pods.Items {
+		pod := &pods.Items[i]
+
+		age := util.Now().Sub(pod.CreationTimestamp.Time)
+		if age.Minutes() < minimumAgeInMinutesToPrune {
+			continue
+		}
+
+		if pod.Status.Phase != kapi.PodRunning && pod.Status.Phase != kapi.PodPending {
+			// not pending or running, skip
+			continue
+		}
+
+		podNode := graph.Pod(g, pod)
+		addPodSpecToGraph(g, &pod.Spec, podNode, registryURL)
+	}
+}
+
+func addPodSpecToGraph(g graph.Graph, spec *kapi.PodSpec, predecessor gonum.Node, registryURL string) {
+	for j := range spec.Containers {
+		container := spec.Containers[j]
+		if !strings.HasPrefix(container.Image, registryURL) {
+			continue
+		}
+		parts := strings.Split(container.Image, "@")
+		if len(parts) != 2 {
+			continue
+		}
+		id := parts[1]
+		if len(id) == 0 {
+			continue
+		}
+		glog.Infof("Looking for node with image id %q", id)
+		imageNode := graph.FindImage(g, id)
+		glog.Infof("Found: %#v", imageNode)
+		if imageNode != nil {
+			g.AddEdge(predecessor, imageNode, graph.ReferencedImageGraphEdgeKind)
+		}
+	}
+}
+
+func addReplicationControllersToGraph(g graph.Graph, rcs *kapi.ReplicationControllerList, registryURL string) {
+	for i := range rcs.Items {
+		rc := &rcs.Items[i]
+		rcNode := graph.ReplicationController(g, rc)
+		addPodSpecToGraph(g, &rc.Spec.Template.Spec, rcNode, registryURL)
+	}
+}
+
+func (f *orphanedImageFinder) Find() []*imageapi.Image {
```

Comment:
Run should be passing enough info to the function that you can do what you need - i.e., maybe you get passed the Node instead of the image and you can look the image up in the graph to find its incoming.  Or you can just pass two arguments to the func - the image, and the image streams that reference it.

----- Original Message -----
> > +		glog.Infof("Found: %#v", imageNode)
> > +		if imageNode != nil {
> > +			g.AddEdge(predecessor, imageNode, graph.ReferencedImageGraphEdgeKind)
> > +		}
> > +	}
> > +}
> > +
> > +func addReplicationControllersToGraph(g graph.Graph, rcs
> > *kapi.ReplicationControllerList, registryURL string) {
> > +	for i := range rcs.Items {
> > +		rc := &rcs.Items[i]
> > +		rcNode := graph.ReplicationController(g, rc)
> > +		addPodSpecToGraph(g, &rc.Spec.Template.Spec, rcNode, registryURL)
> > +	}
> > +}
> > +
> > +func (f *orphanedImageFinder) Find() []*imageapi.Image {
> 
> How about updating the image streams to remove pruned images from
> status.tags... should Run() return the subgraph of pruned images, and
> another function can be used to update the image streams? Or should Run()
> handle that too?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/1841/files#r28822212